File: packages/sdk-python/oshu/task.py
"""
Task handle for managing submitted tasks
"""

import asyncio
from typing import Optional, TypeVar, Generic, Callable, Any, Iterator, Dict
from datetime import datetime

from .types import TaskResult, TaskStatus
from .errors import TaskCancelledError, TaskFailedError, TaskTimeoutError

T = TypeVar('T')


class TaskHandle(Generic[T]):
    """Handle for managing a submitted task"""
    
    def __init__(
        self,
        task_id: str,
        agent_id: str,
        ws_client: Any,  # WebSocketClient
        timeout: float = 600.0,
        session_id: Optional[str] = None,  # SDK session ID if using isolated sandbox
    ):
        self.id = task_id
        self.agent_id = agent_id
        self.session_id = session_id  # SDK session ID for isolated sandbox
        self._ws_client = ws_client
        self._timeout = timeout
        self._status: TaskStatus = 'pending'
        self._result: Optional[TaskResult[T]] = None
        self._listeners: Dict[str, list] = {}
        self._result_future: Optional[asyncio.Future] = None
        
        # Set up message handler
        self._ws_client.on('message', self._handle_message)
    
    def _handle_message(self, message: Dict[str, Any]) -> None:
        """Handle incoming WebSocket message"""
        if message.get('taskId') != self.id:
            return
        
        msg_type = message.get('type', '')
        
        if msg_type == 'task_started':
            self._status = 'processing'
            self._emit('status', 'processing')
        
        elif msg_type == 'stdout':
            self._emit('stdout', message.get('data', ''))
        
        elif msg_type == 'stderr':
            self._emit('stderr', message.get('data', ''))
        
        elif msg_type == 'tool_start':
            self._emit('tool_start', message.get('tool', ''), message.get('input'))
        
        elif msg_type == 'tool_end':
            self._emit('tool_end', message.get('tool', ''), message.get('output'), message.get('duration'))
        
        elif msg_type == 'task_completed':
            self._status = 'completed'
            self._result = TaskResult(
                id=self.id,
                agent_id=self.agent_id,
                status='completed',
                result=message.get('result'),
                output=message.get('output'),
                created_at=datetime.now().isoformat(),
                completed_at=datetime.now().isoformat(),
            )
            self._emit('status', 'completed')
            if self._result_future and not self._result_future.done():
                self._result_future.set_result(self._result)
            self._cleanup()
        
        elif msg_type == 'task_failed':
            self._status = 'failed'
            error = message.get('error', 'Unknown error')
            self._result = TaskResult(
                id=self.id,
                agent_id=self.agent_id,
                status='failed',
                error=error,
                created_at=datetime.now().isoformat(),
                completed_at=datetime.now().isoformat(),
            )
            self._emit('status', 'failed')
            if self._result_future and not self._result_future.done():
                self._result_future.set_exception(TaskFailedError(self.id, error))
            self._cleanup()
        
        elif msg_type == 'task_cancelled':
            self._status = 'cancelled'
            self._result = TaskResult(
                id=self.id,
                agent_id=self.agent_id,
                status='cancelled',
                created_at=datetime.now().isoformat(),
                completed_at=datetime.now().isoformat(),
            )
            self._emit('status', 'cancelled')
            if self._result_future and not self._result_future.done():
                self._result_future.set_exception(TaskCancelledError(self.id))
            self._cleanup()
        
        elif msg_type == 'task_cancelling':
            self._status = 'cancelling'
            self._emit('status', 'cancelling')
        
        elif msg_type == 'task_status':
            status = message.get('status')
            self._status = status
            
            if status == 'completed':
                self._result = TaskResult(
                    id=self.id,
                    agent_id=self.agent_id,
                    status='completed',
                    result=message.get('result'),
                    output=message.get('structuredOutput'),
                    created_at=message.get('createdAt'),
                    started_at=message.get('startedAt'),
                    completed_at=message.get('completedAt'),
                )
                if self._result_future and not self._result_future.done():
                    self._result_future.set_result(self._result)
                self._cleanup()
            elif status == 'failed':
                error = message.get('error', 'Unknown error')
                self._result = TaskResult(
                    id=self.id,
                    agent_id=self.agent_id,
                    status='failed',
                    error=error,
                    created_at=message.get('createdAt'),
                    completed_at=message.get('completedAt'),
                )
                if self._result_future and not self._result_future.done():
                    self._result_future.set_exception(TaskFailedError(self.id, error))
                self._cleanup()
            elif status == 'cancelled':
                self._result = TaskResult(
                    id=self.id,
                    agent_id=self.agent_id,
                    status='cancelled',
                    created_at=message.get('createdAt'),
                    completed_at=message.get('completedAt'),
                )
                if self._result_future and not self._result_future.done():
                    self._result_future.set_exception(TaskCancelledError(self.id))
                self._cleanup()
    
    def _cleanup(self) -> None:
        """Clean up resources"""
        self._ws_client.off('message', self._handle_message)
    
    def on(self, event: str, callback: Callable) -> 'TaskHandle[T]':
        """Register event listener"""
        if event not in self._listeners:
            self._listeners[event] = []
        self._listeners[event].append(callback)
        return self
    
    def _emit(self, event: str, *args) -> None:
        """Emit event to listeners"""
        if event in self._listeners:
            for callback in self._listeners[event]:
                try:
                    callback(*args)
                except Exception:
                    pass
    
    async def cancel(self) -> None:
        """Cancel the task"""
        if self._status in ('completed', 'failed', 'cancelled'):
            return
        await self._ws_client.cancel(self.id)
    
    async def result(self) -> TaskResult[T]:
        """Wait for the task to complete and return the result"""
        # If already have result
        if self._result:
            if self._result.status == 'completed':
                return self._result
            elif self._result.status == 'failed':
                raise TaskFailedError(self.id, self._result.error or 'Unknown error')
            elif self._result.status == 'cancelled':
                raise TaskCancelledError(self.id)
        
        # Create future if not exists
        if not self._result_future:
            self._result_future = asyncio.get_event_loop().create_future()
        
        # Wait with timeout
        try:
            return await asyncio.wait_for(self._result_future, timeout=self._timeout)
        except asyncio.TimeoutError:
            self._cleanup()
            raise TaskTimeoutError(self.id, self._timeout)
    
    def get_status(self) -> TaskStatus:
        """Get current task status"""
        return self._status
    
    def is_finished(self) -> bool:
        """Check if task is finished"""
        return self._status in ('completed', 'failed', 'cancelled')
    
    def stream(self) -> Iterator[Dict[str, Any]]:
        """
        Iterate over task events (synchronous generator).
        Note: For async usage, use event handlers with on() instead.
        """
        import queue
        import threading
        
        event_queue: queue.Queue = queue.Queue()
        finished = threading.Event()
        
        def on_event(msg):
            event_queue.put(msg)
            if msg.get('type') in ('task_completed', 'task_failed', 'task_cancelled'):
                finished.set()
        
        self._ws_client.on('message', on_event)
        
        try:
            while not finished.is_set() or not event_queue.empty():
                try:
                    msg = event_queue.get(timeout=0.1)
                    if msg.get('taskId') == self.id:
                        yield msg
                except queue.Empty:
                    continue
        finally:
            self._ws_client.off('message', on_event)

File: packages/sdk-typescript/src/task.ts
/**
 * Task handle for managing submitted tasks
 */

import { EventEmitter } from 'node:events';
import type { TaskResult, TaskStatus, ServerMessage, TaskEventType, TaskEventHandlers } from './types.js';
import type { WebSocketClient } from './websocket.js';
import { TaskCancelledError, TaskFailedError, TaskTimeoutError } from './errors.js';

export class TaskHandle<T = any> extends EventEmitter {
  public readonly id: string;
  public readonly agentId: string;
  public readonly sessionId?: string;  // SDK session ID if using isolated sandbox
  private wsClient: WebSocketClient;
  private status: TaskStatus = 'pending';
  private _result: TaskResult<T> | null = null;
  private resultPromise: Promise<TaskResult<T>> | null = null;
  private resolveResult: ((result: TaskResult<T>) => void) | null = null;
  private rejectResult: ((error: Error) => void) | null = null;
  private timeout: number;
  private timeoutHandle: NodeJS.Timeout | null = null;

  constructor(
    id: string,
    agentId: string,
    wsClient: WebSocketClient,
    timeout: number = 600, // 10 minutes default in seconds
    sessionId?: string     // Optional SDK session ID for isolated sandbox
  ) {
    super();
    this.id = id;
    this.agentId = agentId;
    this.sessionId = sessionId;
    this.wsClient = wsClient;
    this.timeout = timeout;

    // Set up message handler
    this.wsClient.on('message', this.handleMessage.bind(this));
  }

  private handleMessage(message: ServerMessage) {
    // Only handle messages for this task
    if (!('taskId' in message) || message.taskId !== this.id) {
      return;
    }

    switch (message.type) {
      case 'task_started':
        this.status = 'processing';
        this.emit('status', 'processing');
        break;

      case 'stdout':
        this.emit('stdout', message.data);
        break;

      case 'stderr':
        this.emit('stderr', message.data);
        break;

      case 'tool_start':
        this.emit('tool_start', message.tool, message.input);
        break;

      case 'tool_end':
        this.emit('tool_end', message.tool, message.output, message.duration);
        break;

      case 'task_completed':
        this.status = 'completed';
        this._result = {
          id: this.id,
          agentId: this.agentId,
          status: 'completed',
          result: message.result,
          output: message.output as T,
          createdAt: new Date().toISOString(),
          completedAt: new Date().toISOString(),
        };
        this.emit('status', 'completed');
        this.clearTimeout();
        if (this.resolveResult) {
          this.resolveResult(this._result);
        }
        this.cleanup();
        break;

      case 'task_failed':
        this.status = 'failed';
        this._result = {
          id: this.id,
          agentId: this.agentId,
          status: 'failed',
          error: message.error,
          createdAt: new Date().toISOString(),
          completedAt: new Date().toISOString(),
        };
        this.emit('status', 'failed');
        this.clearTimeout();
        if (this.rejectResult) {
          this.rejectResult(new TaskFailedError(this.id, message.error));
        }
        this.cleanup();
        break;

      case 'task_cancelled':
        this.status = 'cancelled';
        this._result = {
          id: this.id,
          agentId: this.agentId,
          status: 'cancelled',
          createdAt: new Date().toISOString(),
          completedAt: new Date().toISOString(),
        };
        this.emit('status', 'cancelled');
        this.clearTimeout();
        if (this.rejectResult) {
          this.rejectResult(new TaskCancelledError(this.id));
        }
        this.cleanup();
        break;

      case 'task_cancelling':
        this.status = 'cancelling';
        this.emit('status', 'cancelling');
        break;

      case 'task_status':
        // Status update from subscription
        this.status = message.status;
        if (message.status === 'completed') {
          this._result = {
            id: this.id,
            agentId: this.agentId,
            status: 'completed',
            result: message.result,
            output: message.structuredOutput as T,
            createdAt: message.createdAt || new Date().toISOString(),
            startedAt: message.startedAt,
            completedAt: message.completedAt,
          };
          if (this.resolveResult) {
            this.resolveResult(this._result);
          }
          this.cleanup();
        } else if (message.status === 'failed') {
          this._result = {
            id: this.id,
            agentId: this.agentId,
            status: 'failed',
            error: message.error,
            createdAt: message.createdAt || new Date().toISOString(),
            completedAt: message.completedAt,
          };
          if (this.rejectResult) {
            this.rejectResult(new TaskFailedError(this.id, message.error || 'Unknown error'));
          }
          this.cleanup();
        } else if (message.status === 'cancelled') {
          this._result = {
            id: this.id,
            agentId: this.agentId,
            status: 'cancelled',
            createdAt: message.createdAt || new Date().toISOString(),
            completedAt: message.completedAt,
          };
          if (this.rejectResult) {
            this.rejectResult(new TaskCancelledError(this.id));
          }
          this.cleanup();
        }
        break;
    }
  }

  private cleanup() {
    this.wsClient.off('message', this.handleMessage.bind(this));
    this.wsClient.unsubscribe(this.id);
  }

  private clearTimeout() {
    if (this.timeoutHandle) {
      clearTimeout(this.timeoutHandle);
      this.timeoutHandle = null;
    }
  }

  /**
   * Register event handler
   */
  on<K extends TaskEventType>(event: K, handler: TaskEventHandlers[K]): this {
    return super.on(event, handler);
  }

  /**
   * Cancel the task
   */
  async cancel(): Promise<void> {
    if (this.status === 'completed' || this.status === 'failed' || this.status === 'cancelled') {
      return; // Already finished
    }
    this.wsClient.cancel(this.id);
  }

  /**
   * Wait for the task to complete
   */
  async result(): Promise<TaskResult<T>> {
    // If already completed, return cached result
    if (this._result) {
      if (this._result.status === 'completed') {
        return this._result;
      } else if (this._result.status === 'failed') {
        throw new TaskFailedError(this.id, this._result.error || 'Unknown error');
      } else if (this._result.status === 'cancelled') {
        throw new TaskCancelledError(this.id);
      }
    }

    // Create promise if not already waiting
    if (!this.resultPromise) {
      this.resultPromise = new Promise<TaskResult<T>>((resolve, reject) => {
        this.resolveResult = resolve;
        this.rejectResult = reject;

        // Set timeout (convert seconds to milliseconds)
        this.timeoutHandle = setTimeout(() => {
          reject(new TaskTimeoutError(this.id, this.timeout));
          this.cleanup();
        }, this.timeout * 1000);
      });
    }

    return this.resultPromise;
  }

  /**
   * Get current status
   */
  getStatus(): TaskStatus {
    return this.status;
  }

  /**
   * Check if task is finished
   */
  isFinished(): boolean {
    return this.status === 'completed' || this.status === 'failed' || this.status === 'cancelled';
  }
}
